Sources/StreamChat/Workers/Background/ConnectionRecoveryUpdater.swift

```diff
@@ -101,81 +101,141 @@ class ConnectionRecoveryUpdater: EventWorker {
     }
     
     private func fetchAndReplayMissingEvents() {
-        database.backgroundReadOnlyContext.perform { [weak self, useSyncEndpoint] in
-            let refetchExistingQueries: () -> Void = {
-                self?.databaseCleanupUpdater.refetchExistingChannelListQueries()
-            }
-            
-            if useSyncEndpoint {
-                self?.sync(completion: refetchExistingQueries)
-            } else {
-                refetchExistingQueries()
+        guard let lastSyncedAt = lastSyncedAt else { return }
+        
+        log.debug("Will start sync starting from \(lastSyncedAt)")
+        
+        if useSyncEndpoint {
+            fetchLocalChannelIDs { [weak self] cids in
+                self?.getMissingEvents(for: cids, since: lastSyncedAt) { error in
+                    self?.syncChannelListQueries(syncedChannelIDs: error == nil ? cids : [])
+                }
             }
+        } else {
+            syncChannelListQueries(syncedChannelIDs: [])
         }
     }
-
-    private func sync(completion: @escaping () -> Void) {
-        guard let lastSyncedAt = lastSyncedAt else { return }
+    
+    private func fetchFirstPage(
+        of queries: [ChannelListQuery],
+        completion: @escaping ([String: Result<ChannelListPayload, Error>]) -> Void
+    ) {
+        let group = DispatchGroup()
+        let semaphore = DispatchSemaphore(value: 1)
```

Comment:
Why do we need both a group and a semaphore? The group should be enough 🤔

Reply:
A group is for tracking when all tasks are done, a `semaphore` makes sure we don't write `results` dict from multiple threads at the same time 👍

Reply:
Ah I see, makes sense 👍 Thanks!